File: frontend/src/api/series.ts
import { apiFetch } from "./client";
import type { Game, LeaderboardRow, PaginationInfo, Series, User, UUID } from "../types/dto";

export interface CreateSeriesRequest {
  name: string;
  description: string;
  start_at: string;
  end_at: string;
  price_rub: number;
  is_closed: boolean;
  game_type: number;
}

export function createSeries(req: CreateSeriesRequest) {
  return apiFetch<Series>("/api/v1/series", { method: "POST", body: JSON.stringify(req) });
}

export function getSeries(id: UUID) {
  return apiFetch<Series>(`/api/v1/series/${id}`, { method: "GET" });
}

export interface GetSeriesFullResponse {
  series: Series;
  participants: GetSeriesParticipantsResponse;
  games: {
    items: Game[];
    pagination: PaginationInfo;
  };
  leaderboard: GetSeriesLeaderboardResponse;
}

export function getSeriesFull(
  id: UUID,
  params: {
    participants_limit?: number;
    participants_offset?: number;
    games_limit?: number;
    games_offset?: number;
    leaderboard_limit?: number;
    leaderboard_offset?: number;
  } = {}
) {
  const q = new URLSearchParams();
  if (params.participants_limit != null) q.set("participants_limit", String(params.participants_limit));
  if (params.participants_offset != null) q.set("participants_offset", String(params.participants_offset));
  if (params.games_limit != null) q.set("games_limit", String(params.games_limit));
  if (params.games_offset != null) q.set("games_offset", String(params.games_offset));
  if (params.leaderboard_limit != null) q.set("leaderboard_limit", String(params.leaderboard_limit));
  if (params.leaderboard_offset != null) q.set("leaderboard_offset", String(params.leaderboard_offset));
  return apiFetch<GetSeriesFullResponse>(`/api/v1/series/${id}/full?${q.toString()}`, { method: "GET" });
}

export function updateSeries(id: UUID, patch: Partial<CreateSeriesRequest>) {
  return apiFetch<Series>(`/api/v1/series/${id}`, { method: "PATCH", body: JSON.stringify(patch) });
}

export function deleteSeries(id: UUID) {
  return apiFetch<void>(`/api/v1/series/${id}`, { method: "DELETE" });
}

export interface GetClubSeriesResponse {
  items: Series[];
  pagination: PaginationInfo;
}

export function getClubSeries(clubId: UUID, params: { limit?: number; offset?: number }) {
  const q = new URLSearchParams();
  if (params.limit != null) q.set("limit", String(params.limit));
  if (params.offset != null) q.set("offset", String(params.offset));
  return apiFetch<GetClubSeriesResponse>(`/api/v1/club/${clubId}/series?${q.toString()}`, { method: "GET" });
}

export interface AllSeriesItem {
  id: UUID;
  club_id: UUID;
  club_name: string;
  name: string;
  description: string;
  start_at: string;
  end_at: string;
  games_count: number;
}

export interface GetAllSeriesResponse {
  items: AllSeriesItem[];
  pagination: PaginationInfo;
}

export function listAllSeries(params: { limit?: number; offset?: number }) {
  const q = new URLSearchParams();
  if (params.limit != null) q.set("limit", String(params.limit));
  if (params.offset != null) q.set("offset", String(params.offset));
  return apiFetch<GetAllSeriesResponse>(`/api/v1/series/all?${q.toString()}`, { method: "GET" });
}

export function joinSeries(id: UUID) {
  return apiFetch<void>(`/api/v1/series/${id}/join`, { method: "POST" });
}

export function leaveSeries(id: UUID) {
  return apiFetch<void>(`/api/v1/series/${id}/leave`, { method: "POST" });
}

export interface GetSeriesParticipantsResponse {
  items: User[];
  pagination: PaginationInfo;
}

export function getSeriesParticipants(id: UUID, params: { limit?: number; offset?: number; q?: string }) {
  const q = new URLSearchParams();
  if (params.limit != null) q.set("limit", String(params.limit));
  if (params.offset != null) q.set("offset", String(params.offset));
  if (params.q) q.set("q", params.q);
  return apiFetch<GetSeriesParticipantsResponse>(`/api/v1/series/${id}/participants?${q.toString()}`, { method: "GET" });
}

export interface GetSeriesLeaderboardResponse {
  items: LeaderboardRow[];
  pagination: PaginationInfo;
}

export function getLeaderboard(id: UUID, params: { limit?: number; offset?: number }) {
  const q = new URLSearchParams();
  if (params.limit != null) q.set("limit", String(params.limit));
  if (params.offset != null) q.set("offset", String(params.offset));
  return apiFetch<GetSeriesLeaderboardResponse>(`/api/v1/series/${id}/leaderboard?${q.toString()}`, { method: "GET" });
}

File: backend/internal/domain/dto/series.go
package dto

import (
	"SmartLeague/internal/domain/types"
	"time"

	"github.com/google/uuid"
)

type Series struct {
	ID          uuid.UUID          `json:"id"`
	ClubID      uuid.UUID          `json:"club_id"`
	CreatorID   *uuid.UUID         `json:"creator_id,omitempty"`
	Name        string             `json:"name"`
	Description string             `json:"description"`
	StartAt     time.Time          `json:"start_at"`
	EndAt       time.Time          `json:"end_at"`
	PriceRub    int                `json:"price_rub"`
	IsClosed    bool               `json:"is_closed"`
	GameType    types.GameType     `json:"game_type"`
	Status      types.SeriesStatus `json:"status"`
}

type CreateSeriesRequest struct {
	Name        string             `json:"name" validate:"required,min=1,max=200"`
	Description string             `json:"description" validate:"required,min=1,max=10000"`
	StartAt     time.Time          `json:"start_at" validate:"required"`
	EndAt       time.Time          `json:"end_at" validate:"required"`
	PriceRub    int                `json:"price_rub" validate:"min=0,max=100000000"`
	IsClosed    bool               `json:"is_closed"`
	GameType    types.GameType     `json:"game_type" validate:"eq=0"`
	Status      types.SeriesStatus `json:"status" validate:"min=0,max=3"`
}

type CreateSeriesResponse Series

type GetSeriesRequest struct {
	ID uuid.UUID `json:"-" validate:"required,uuid" swaggerignore:"true"`
}

type GetSeriesResponse Series

type GetSeriesFullRequest struct {
	ID                 uuid.UUID `json:"-" validate:"required,uuid" swaggerignore:"true"`
	ParticipantsLimit  *int      `json:"participants_limit,omitempty" form:"participants_limit" validate:"omitempty,min=1,max=100"`
	ParticipantsOffset *int      `json:"participants_offset,omitempty" form:"participants_offset" validate:"omitempty,min=0"`
	GamesLimit         *int      `json:"games_limit,omitempty" form:"games_limit" validate:"omitempty,min=1,max=100"`
	GamesOffset        *int      `json:"games_offset,omitempty" form:"games_offset" validate:"omitempty,min=0"`
	LeaderboardLimit   *int      `json:"leaderboard_limit,omitempty" form:"leaderboard_limit" validate:"omitempty,min=1,max=100"`
	LeaderboardOffset  *int      `json:"leaderboard_offset,omitempty" form:"leaderboard_offset" validate:"omitempty,min=0"`
}

type GetSeriesFullResponse struct {
	Series       *Series                        `json:"series"`
	Participants *GetSeriesParticipantsResponse `json:"participants"`
	Games        *GetSeriesGamesResponse        `json:"games"`
	Leaderboard  *GetSeriesLeaderboardResponse  `json:"leaderboard"`
}

type GetClubSeriesRequest struct {
	ClubID uuid.UUID `json:"-" validate:"required,uuid" swaggerignore:"true"`
	Limit  *int      `json:"limit,omitempty" form:"limit" validate:"omitempty,min=1,max=100"`
	Offset *int      `json:"offset,omitempty" form:"offset" validate:"omitempty,min=0"`
}

type GetClubSeriesResponse struct {
	Items      []*Series      `json:"items"`
	Pagination PaginationInfo `json:"pagination"`
}

type AllSeriesItem struct {
	ID          uuid.UUID `json:"id"`
	ClubID      uuid.UUID `json:"club_id"`
	ClubName    string    `json:"club_name"`
	Name        string    `json:"name"`
	Description string    `json:"description"`
	StartAt     time.Time `json:"start_at"`
	EndAt       time.Time `json:"end_at"`
	GamesCount  int       `json:"games_count"`
}

type GetAllSeriesRequest struct {
	Limit  *int `json:"limit,omitempty" form:"limit" validate:"omitempty,min=1,max=100"`
	Offset *int `json:"offset,omitempty" form:"offset" validate:"omitempty,min=0"`
}

type GetAllSeriesResponse struct {
	Items      []*AllSeriesItem `json:"items"`
	Pagination PaginationInfo   `json:"pagination"`
}

type UpdateSeriesRequest struct {
	ID          uuid.UUID           `json:"-" validate:"required,uuid" swaggerignore:"true"`
	Name        *string             `json:"name,omitempty" validate:"omitempty,min=1,max=200"`
	Description *string             `json:"description,omitempty" validate:"omitempty,min=1,max=10000"`
	StartAt     *time.Time          `json:"start_at,omitempty"`
	EndAt       *time.Time          `json:"end_at,omitempty"`
	PriceRub    *int                `json:"price_rub,omitempty" validate:"omitempty,min=0,max=100000000"`
	IsClosed    *bool               `json:"is_closed,omitempty"`
	GameType    *types.GameType     `json:"game_type,omitempty" validate:"omitempty,eq=0"`
	Status      *types.SeriesStatus `json:"status,omitempty" validate:"omitempty,min=0,max=3"`
}

type UpdateSeriesResponse Series

type DeleteSeriesRequest struct {
	ID uuid.UUID `json:"-" validate:"required,uuid" swaggerignore:"true"`
}

type GetSeriesParticipantsRequest struct {
	SeriesID uuid.UUID `json:"-" validate:"required,uuid" swaggerignore:"true"`
	Limit    *int      `json:"limit,omitempty" form:"limit" validate:"omitempty,min=1,max=100"`
	Offset   *int      `json:"offset,omitempty" form:"offset" validate:"omitempty,min=0"`
	Query    *string   `json:"q,omitempty" form:"q" validate:"omitempty,min=1,max=100"`
}

type GetSeriesParticipantsResponse struct {
	Items      []*User        `json:"items"`
	Pagination PaginationInfo `json:"pagination"`
}

type JoinSeriesRequest struct {
	SeriesID  uuid.UUID `json:"-" validate:"required,uuid" swaggerignore:"true"`
	ProfileID uuid.UUID `json:"-" validate:"required,uuid" swaggerignore:"true"`
}

type LeaveSeriesRequest struct {
	SeriesID  uuid.UUID `json:"-" validate:"required,uuid" swaggerignore:"true"`
	ProfileID uuid.UUID `json:"-" validate:"required,uuid" swaggerignore:"true"`
}
